include/xtensor/xstorage.hpp

```diff
@@ -1169,6 +1192,116 @@ namespace xt
+#endif
+
+    template <class T, std::size_t N, std::size_t Align = SELECT_ALIGN>
+    class alignas(Align) aligned_array : public std::array<T, N>
+    {
+    public:
+        // Note: this is for alignment detection. The allocator serves no other purpose than
+        //       that of a trait here.
+        using allocator_type = std::conditional_t<Align != 0,
+                                                  xsimd::aligned_allocator<T, Align>,
+                                                  std::allocator<T>>;
+    };
+
+    /**
+     * This array class is modeled after ``std::array`` but adds optional alignment through a template
+     * parameter and all accessors/iterators (except reverse) are constexpr qualified.
+     *
+     * To be moved to xtl, along with the rest of xstorage.hpp
+     */
+    template <class T, std::size_t N>
+    struct const_array
+    {
+        using size_type = std::size_t;
+        using value_type = T;
+        using pointer = value_type*;
+        using const_pointer = const value_type*;
+        using reference = value_type&;
+        using const_reference = const value_type&;
+        using difference_type = std::ptrdiff_t;
+        using iterator = pointer;
+        using const_iterator = const_pointer;
+
+        using reverse_iterator = std::reverse_iterator<const_iterator>;
+        using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+        constexpr const_reference operator[](std::size_t idx) const
+        {
+            return m_data[idx];
+        }
+
+        constexpr const_iterator begin() const noexcept
+        {
+            return cbegin();
+        }
+
+        constexpr const_iterator end() const noexcept
+        {
+            return cend();
+        }
+
+        constexpr const_iterator cbegin() const noexcept
+        {
+            return m_data;
+        }
+
+        constexpr const_iterator cend() const noexcept
+        {
+            return m_data + N;
+        }
+
+        // TODO make constexpr once C++17 arrives
+        reverse_iterator rbegin() const noexcept
+        {
+            return crbegin();
+        }
+
+        reverse_iterator rend() const noexcept
+        {
+            return crend();
+        }
+
+        const_reverse_iterator crbegin() const noexcept
+        {
+            return const_reverse_iterator(end());
+        }
+
+        const_reverse_iterator crend() const noexcept
+        {
+            return const_reverse_iterator(begin());
+        }
+
+        constexpr const_pointer data() const noexcept
+        {
+            return m_data;
+        }
+
+        constexpr const_reference front() const
+        {
+            return m_data[0];
+        }
+
+        constexpr const_reference back() const
+        {
+            return m_data[size() - 1];
+        }
+
+        constexpr size_type size() const
+        {
+            return N;
+        }
+
+        T m_data[N > 0 ? N : 1];
```

Comment:
It looks like the `const` has been lost when the code was moved.

Reply:
thanks for spotting.